Adds react-native support via Expo

stigi opened the pull request:
Hi there,

I've seen the `react-native` branch and thought I'd make it work.

For RN we need both CLJS compilation (via shadow) and JS and asset bundling which is done via Expo. I've used [concurrently](https://www.npmjs.com/package/concurrently) to run both of them in parallel via `yarn dev`. STDIN is passed to Expo, so that one can open the iOS simulator or Android emulator by pressing `i` or `a`.

~I wasn't super sure about which hot reloading / fast-refresh approach to use, so I've opted with the `shadow.expo` built in, which required to disable react natives fast refresh. I've added some local state (via `counter`) to the app to demonstrate how nicely shadows build in reloading works while keeping local state.~

Update on fast refresh: The PR now uses UIx fast-refresh, instead of the implementation in `shadow.expo`.

There's still some things buggy in `create-uix-app` and I couldn't find where the repo that CLI is actually hosted, but the `uix-starter` branch should work now.

**Demo**

https://github.com/pitch-io/uix-starter/assets/13815/fba7ae85-b638-4e93-801b-a87fc61aeb5d




**Future work:**
~It's trivial to add [react-native-web](https://necolas.github.io/react-native-web/) support as well. Let me know if there's interest for this.~

Update: react-native-web support has been added to this PR.

stigi committed:
Adding expo

stigi committed:
Starting expo in parallel to shadow-cljs via concurrently

stigi committed:
Adding counter with local state

to demonstrate hot reload

stigi commented on app.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "expo": {
+    "name": "uix-starter",
+    "slug": "uix-starter",
```
Maybe `create-uix-app` can replace the project name with the one provided.

stigi commented on package.json:
```diff
@@ -1,9 +1,19 @@
+    "expo": "~49.0.13",
+    "react": "18.2.0",
+    "react-native": "0.72.5",
+    "create-react-class": "^15.7.0"
```
Shadows Expo wrapper uses `create-react-class`

https://github.com/thheller/shadow-cljs/blob/77b5dbef335d4d6f96fafaeb6658aab9bed17f1e/src/main/shadow/expo.cljs#L1-L4

roman01la commented on shadow-cljs.edn:
```diff
@@ -1,10 +1,11 @@
-           :devtools {:preloads [preload]
-                      :build-notify preload/build-notify}}}}
\ No newline at end of file
+           :devtools {:autoload true
```
what's `:autoload`?

stigi replied:
> [`:autoload`](https://shadow-cljs.github.io/docs/UsersGuide.html#_lifecycle_hooks)
> A boolean controlling whether code should be hot loaded. Implicitly set to true if either of the callbacks is set. Always enabled for the :browser target by default, set to false to disable.

I took it from https://github.com/thheller/reagent-expo/blob/5886b98f1fbe5c173ede24520d72a76cebac2b98/shadow-cljs.edn#L15

roman01la commented:
@stigi Thanks! The RN template is already used in `create-uix-app`, see RN docs in UIx repo https://github.com/pitch-io/uix/blob/master/docs/react-native.md

Here's the repo for the CLI script https://github.com/roman01la/create-uix-app

We can add a template specifically for Expo though, can I ask you to open this PR against a different branch? Maybe `react-native-expo`

stigi commented:
@roman01la I wasn't properly reading the docs (see #7). Now I understand that the existing `--react-native` template was intended to add UIX and CLJS to an existing react-native app.

Intuitively I was under the impression `create-uix-app` would always create a new app from scratch, and this is what this PR is trying to provide. I think it makes perfect sense to have a `create-uix-app --expo` template. That could then include react-native and react-native-web. It's a super powerful demo.

I would be keen on getting your feedback on the hot-reloading approach to understand if there are better ways to do it.

roman01la commented:
> I would be keen on getting your feedback on the hot-reloading approach to understand if there are better ways to do it.

I'm not sure how hot-reloading works in Expo, but for bare RN project it's recommended to disable RN's fast refresh and use UIx react refresh integration, this is also described in the docs here https://github.com/pitch-io/uix/blob/master/docs/react-native.md

stigi commented:
> I'm not sure how hot-reloading works in Expo, but for bare RN project it's recommended to disable RN's fast refresh and use UIx react refresh integration, this is also described in the docs here https://github.com/pitch-io/uix/blob/master/docs/react-native.md

Expo is using default react-native fast refresh. In this PR I used shadow-cljs proposed way to reload expo apps. It's implemented in [`shadow.expo`](https://github.com/thheller/shadow-cljs/blob/77b5dbef335d4d6f96fafaeb6658aab9bed17f1e/src/main/shadow/expo.cljs). I could alternatively switch over to the UIx react refresh integration to see if that performs similarly well. Let me give it a shot.

roman01la commented:
looks like shadow.expo re-renders from the root, which how Reagent hot-reloading is traditionally done as well, for UIx it's not needed, since react-refresh takes care of swapping components and preserving hooks state

stigi committed:
Using plain UIx instead of `shadow.expo` to register expo root & do refresh

stigi commented:
I've removed the usage of `shadow.expo` and am directly registering the root component with expo in 20d86e5.

Refresh seems to work nicely, keeping local state intact. It's also nicely fast :)

stigi committed:
Removing unused crc dep

stigi committed:
Adding react-native-web dependencies

stigi committed:
Updats hot-reload disabling for rn-web

stigi committed:
Adds missing peer dependencies

stigi committed:
Using modern `Pressable` over `Button`

stigi commented:
@roman01la I've did some updates to the PR and am pretty happy with it's shape now. 

I've removed `shadow.expo` and added support for web via `react-native-web`. It's super cool 🤩 Please check it out if you have time.

Simply run `yarn dev` and then press `w` to bring up web.

I'd love to see this integrated as `create-uix-app --expo`. Let me know how I can help with that.


https://github.com/pitch-io/uix-starter/assets/13815/9f035bf1-36f8-413a-979d-3c44ca199da6

stigi committed:
Moving `setHotLoadingEnabled` to `preload.cljs`

Also addresses a bug on Android where calling it in code would cause an infinite reloading loop

roman01la commented:
@stigi When I press `w` I get the following error in my terminal
```
Starting Webpack on port 19006 in development mode.
ValidationError: Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
 - configuration.entry should be a non-empty array.
   -> All modules are loaded upon startup. The last one is exported.
ValidationError: Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
 - configuration.entry should be a non-empty array.
   -> All modules are loaded upon startup. The last one is exported.
```

roman01la commented:
iOS build is working great 👍

roman01la commented:
Works great now, thanks!
